hivemind/utils/mpfuture.py

```diff
@@ -2,171 +2,198 @@
 
 import asyncio
 import concurrent.futures._base as base
+import ctypes
 import multiprocessing as mp
 import multiprocessing.connection
-import time
-from functools import lru_cache
-from typing import Optional, Tuple, Generic, TypeVar
+import os
+import threading
+from typing import Tuple, Generic, TypeVar, Dict, Optional, Any, Callable
+from hivemind.utils.logging import get_logger
 
-from hivemind.utils.threading import run_in_background
 
-ResultType = TypeVar('ResultType')
+logger = get_logger(__name__)
 
+# flavour types
+ResultType = TypeVar('ResultType')
+PID, UID, State, PipeEnd = int, int, Any, mp.connection.Connection
+ALL_STATES = base.PENDING, base.RUNNING, base.FINISHED, base.CANCELLED, base.CANCELLED_AND_NOTIFIED
+TERMINAL_STATES = {base.FINISHED, base.CANCELLED, base.CANCELLED_AND_NOTIFIED}
+
+INITIALIZER_LOCK = mp.Lock()
+PIPE_WAITER: Optional[threading.Thread] = None
+MPFUTURE_PIPES: Dict[PID, Tuple[PipeEnd, PipeEnd]] = mp.Manager().dict()
+ACTIVE_FUTURES: Optional[Dict[PID, MPFuture]] = None
+ACTIVE_PID: Optional[PID] = None
```

Comment:
The entities in this block (from `INITIALIZER_LOCK` to `ACTIVE_PID`) are global variables, not constants. Their names should not be capitalized (see [this](https://www.python.org/dev/peps/pep-0008/#id44) PEP-8 section).

I would also suggest to make them private static members of the `MPFuture` class along with the `_initialize_mpfuture_backend`, `_send_update`, and `_process_updates_in_background` methods below. This is because the single purpose of them is to make the `MPFuture` instances work.

Reply:
[implementing as suggested]